### PR TITLE
Add `distance` expression support

### PR DESCRIPTION
**Pull request by HarelM**
## Launch Checklist

This allows the ability to use `distance` expression in web. 

- https://github.com/maplibre/maplibre-gl-js/issues/2077

This is heavily based on the c++ native implementation:
https://github.com/maplibre/maplibre-native/blob/cd6a0ab11fcd9b5bbfee87896a14159d0ed56256/src/mbgl/style/expression/distance.cpp
https://github.com/maplibre/maplibre-native/blob/cd6a0ab11fcd9b5bbfee87896a14159d0ed56256/src/mbgl/tile/geometry_tile_data.cpp
https://github.com/maplibre/maplibre-native/blob/cd6a0ab11fcd9b5bbfee87896a14159d0ed56256/src/mbgl/util/geometry_util.cpp

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by HarelM** [comment]
FYI: @louwers, thanks for the pointers!

File: src/index.ts [hunk 1]
```diff
@@ -187,6 +188,7 @@ export {
     format,
     validate,
     migrate,
+    classifyRings,
```
Comment (HarelM) [inline src/index.ts:191]: This is exported here in order to remove it from maplibre-gl code base.

**Comment by HarelM** [comment]
I'm happy with the tests coverage and the implementation, it found some bugs I needed to fix from wrong conversion of code and differences between c++ and javascript (integer division).
Overall the tests cover the scenarios that I wanted to check (points, polygons with holes, lines etc).
This is good to go from my point of view.

**Comment by HarelM** [comment]
Also @louwers I think the following line should use the point constant and not the line constant:
https://github.com/maplibre/maplibre-native/blob/cd6a0ab11fcd9b5bbfee87896a14159d0ed56256/src/mbgl/style/expression/distance.cpp#L373

I don't think it will dramatically affect the algorithm, but when I looked at the code and converted it, I came across this.

**Review by neodescis — APPROVED** [review]
I'm not familiar enough with the math to check it, but it looks good to me in general. Certainly great job on improving typings, and the tests look comprehensive.

**Comment by HarelM** [comment]
This adds about 7k unzipped to the bundle size (~1%). It's not negligible but it's part of the spec, so I don't see a way around this.

**Pull request merged**